pkg/services/ngalert/models/alert_rule.go

```diff
@@ -0,0 +1,179 @@
+)
+
+type NoDataState string
+
+func (noDataState NoDataState) String() string {
+	return string(noDataState)
+}
+
+const (
+	Alerting      NoDataState = "Alerting"
+	NoData        NoDataState = "NoData"
+	KeepLastState NoDataState = "KeepLastState"
+	OK            NoDataState = "OK"
+)
+
+type ExecutionErrorState string
+
+func (executionErrorState ExecutionErrorState) String() string {
+	return string(executionErrorState)
+}
+
+const (
+	AlertingErrState      ExecutionErrorState = "Alerting"
+	KeepLastStateErrState ExecutionErrorState = "KeepLastState"
+)
+
+// AlertRule is the model for alert rules in unified alerting.
+type AlertRule struct {
+	ID              int64 `xorm:"pk autoincr 'id'"`
+	OrgID           int64 `xorm:"org_id"`
+	Title           string
+	Condition       string
+	Data            []AlertQuery
+	Updated         time.Time
+	IntervalSeconds int64
+	Version         int64
+	UID             string `xorm:"uid"`
+	NamespaceUID    string `xorm:"namespace_uid"`
+	RuleGroup       string
+	NoDataState     NoDataState
+	ExecErrState    ExecutionErrorState
+	// Receivers are used for migrating notification channels of existing alerts
+	// Receivers []string
+}
+
+// AlertRuleKey is the alert definition identifier
+type AlertRuleKey struct {
+	OrgID int64
+	UID   string
+}
+
+func (k AlertRuleKey) String() string {
+	return fmt.Sprintf("{orgID: %d, UID: %s}", k.OrgID, k.UID)
+}
+
+// GetKey returns the alert definitions identifier
+func (alertRule *AlertRule) GetKey() AlertRuleKey {
+	return AlertRuleKey{OrgID: alertRule.OrgID, UID: alertRule.UID}
+}
+
+// PreSave sets default values and loads the updated model for each alert query.
+func (alertRule *AlertRule) PreSave(timeNow func() time.Time) error {
+	for i, q := range alertRule.Data {
+		err := q.PreSave()
+		if err != nil {
+			return fmt.Errorf("invalid alert query %s: %w", q.RefID, err)
+		}
+		alertRule.Data[i] = q
+	}
+	alertRule.Updated = timeNow()
+	return nil
+}
+
+// AlertRuleVersion is the model for alert rule versions in unified alerting.
+type AlertRuleVersion struct {
+	ID               int64  `xorm:"pk autoincr 'id'"`
+	RuleOrgID        int64  `xorm:"rule_org_id"`
+	RuleUID          string `xorm:"rule_uid"`
+	RuleNamespaceUID string `xorm:"rule_namespace_uid"`
+	RuleGroup        string
+	ParentVersion    int64
+	RestoredFrom     int64
+	Version          int64
+
+	Created         time.Time
+	Title           string
+	Condition       string
+	Data            []AlertQuery
+	IntervalSeconds int64
+	NoDataState     NoDataState
+	ExecErrState    ExecutionErrorState
+	// Receivers are used for migrating notification channels of existing alerts
+	// do we still need this?
+	// Receivers []string
+}
+
+// GetAlertRuleByUIDQuery is the query for retrieving/deleting an alert rule by UID and organisation ID.
+type GetAlertRuleByUIDQuery struct {
+	UID   string
+	OrgID int64
+
+	Result *AlertRule
+}
+
+// DeleteAlertRuleByUIDCommand is the command for deleting an alert rule
```

Comment:
Is there any reason to use *Command structs to package up parameters, instead of passing them directly to the function(s)? The command pattern is tied to the bus, but as the bus is falling out of use, I'm unsure why the command pattern is still relevant.